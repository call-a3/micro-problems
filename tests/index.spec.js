import declareProblem from '../src/index'

const problemDescription = {
  title: 'Something went wrong',
  status: 500,
  detail: 'There is nothing you can do',
}

describe('micro-problems', () => {
  it('should return an Error subclass', () => {
    const { Problem } = declareProblem(problemDescription)
    expect(new Problem()).toBeInstanceOf(Error)
  })

  it('should return an HTTP handler', () => {
    const { handler } = declareProblem(problemDescription)
    const res = {
      setHeader: jest.fn(),
      getHeader: jest.fn(() => true),
      end: jest.fn(),
    }
    handler(null, res)
    expect(res).toHaveProperty('statusCode', 200)
    expect(res.setHeader).toHaveBeenCalled()
    expect(res.setHeader.mock.calls).toMatchSnapshot('setHeader calls')
    expect(res.end).toHaveBeenCalledTimes(1)
    expect(res.end.mock.calls[0]).toMatchSnapshot('body')
  })

  describe('decorator', () => {
    const { Problem, decorator } = declareProblem(problemDescription)
    it('should call the wrapped handler', async () => {
      const fn = jest.fn()
      const decorated = decorator(fn)
      await decorated(null, null)
      expect(fn).toHaveBeenCalled()
    })
    it('should catch the Problem thrown from the wrapped handler', async () => {
      const fn = jest.fn(() => {
        throw new Problem('This instance could not be helped', {
          meta: 'void',
        })
      })
      const res = {
        setHeader: jest.fn(),
        getHeader: jest.fn(() => true),
        end: jest.fn(),
      }
      const decorated = decorator(fn)
      await decorated(null, res)
      expect(res).toHaveProperty('statusCode', problemDescription.status)
      expect(res.setHeader).toHaveBeenCalled()
      expect(res.setHeader.mock.calls).toMatchSnapshot('setHeader calls')
      expect(res.end).toHaveBeenCalledTimes(1)
      expect(res.end.mock.calls[0]).toMatchSnapshot('body')
    })
    it('should throw other Error thrown from the wrapped handler', async () => {
      const fn = jest.fn(() => {
        throw new Error('generic error')
      })
      const res = {
        setHeader: jest.fn(),
        getHeader: jest.fn(() => true),
        end: jest.fn(),
      }
      const decorated = decorator(fn)
      await expect(decorated(null, res)).rejects.toThrowErrorMatchingSnapshot()
    })
  })
})
