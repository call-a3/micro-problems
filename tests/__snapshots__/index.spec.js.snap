// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`micro-problems decorator should catch the Problem thrown from the wrapped handler: body 1`] = `
Array [
  "{\\"title\\":\\"Something went wrong\\",\\"status\\":500,\\"detail\\":\\"This instance could not be helped\\",\\"meta\\":\\"void\\"}",
]
`;

exports[`micro-problems decorator should catch the Problem thrown from the wrapped handler: setHeader calls 1`] = `
Array [
  Array [
    "Content-Type",
    "application/json+problem",
  ],
  Array [
    "Content-Length",
    104,
  ],
]
`;

exports[`micro-problems decorator should throw other Error thrown from the wrapped handler 1`] = `"generic error"`;

exports[`micro-problems should return an HTTP handler: body 1`] = `
Array [
  "{\\"title\\":\\"Something went wrong\\",\\"status\\":500,\\"detail\\":\\"There is nothing you can do\\"}",
]
`;

exports[`micro-problems should return an HTTP handler: setHeader calls 1`] = `
Array [
  Array [
    "Content-Length",
    84,
  ],
]
`;
